swap/swap.go

```diff
@@ -17,69 +17,340 @@
 package swap
 
 import (
+	"context"
+	"crypto/ecdsa"
 	"errors"
 	"fmt"
+	"math/big"
 	"strconv"
 	"sync"
+	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/contracts/swap"
+	cswap "github.com/ethersphere/swarm/contracts/swap"
 	"github.com/ethersphere/swarm/log"
 	"github.com/ethersphere/swarm/p2p/protocols"
 	"github.com/ethersphere/swarm/state"
 )
 
-// SwAP Swarm Accounting Protocol
+// ErrInvalidChequeSignature indicates the signature on the cheque was invalid
+var ErrInvalidChequeSignature = errors.New("invalid cheque signature")
+
+// Swap represents the SwAP Swarm Accounting Protocol
 // a peer to peer micropayment system
 // A node maintains an individual balance with every peer
 // Only messages which have a price will be accounted for
 type Swap struct {
-	stateStore state.Store        //stateStore is needed in order to keep balances across sessions
-	lock       sync.RWMutex       //lock the balances
-	balances   map[enode.ID]int64 //map of balances for each peer
+	api                 PublicAPI
+	stateStore          state.Store          // stateStore is needed in order to keep balances across sessions
+	lock                sync.RWMutex         // lock the balances
+	balances            map[enode.ID]int64   // map of balances for each peer
+	cheques             map[enode.ID]*Cheque // map of balances for each peer
+	peers               map[enode.ID]*Peer   // map of all swap Peers
+	backend             cswap.Backend        // the backend (blockchain) used
+	owner               *Owner               // contract access
+	params              *Params              // economic and operational parameters
+	contractReference   *swap.Swap           // reference to the smart contract
+	oracle              PriceOracle          // the oracle providing the ether price for honey
+	paymentThreshold    int64                // balance difference required for requesting cheque
+	disconnectThreshold int64                // balance difference required for dropping peer
+}
+
+// Owner encapsulates information related to accessing the contract
+type Owner struct {
+	Contract   common.Address    // address of swap contract
+	address    common.Address    // owner address
+	privateKey *ecdsa.PrivateKey // private key
```

Comment:
not sure about passing around private keys

Reply:
https://github.com/ethersphere/swarm/pull/1554/files#r308376714